design_notes/v2.md

```diff
@@ -0,0 +1,127 @@
+
+## Definitions
+
+- `Parser` - A parser is a function that constructs a `ManifestStore` from an `AsyncObjectReader` and a `file_url`.
+
+## Protocols
+
+**Note**: We anticipate using [obspec](https://developmentseed.org/obspec/latest/) with the following typing in the future, but currently use only `ObjectStore` rather than the general interface definition.
+
+```python
+from obspec import GetAsync, GetRangeAsync, List
+
+class AsyncObjectReader(GetAsync, GetRangeAsync, List):
+    ...
+
+```
+
+```python
+class Parser(protocol):
+    """
+    Callback protocol for creating a virtual zarr store from a file
+    """
+    def __call__(
+        self,
+        file_url: str,
+        object_reader: ObjectStore,
+        *,
+    ) → ManifestStore:
+        ...
+```
+
+## Top-level API functions
+
+- `open_virtual_dataset` - construct a Xarray Dataset with NumPy and/or ManifestArray-backed variables from a file
+- `open_virtual_mfdataset` - construct a Xarray Dataset with NumPy and/or ManifestArray-backed variables from many files
+
+> **NOTE**: Here the construction of a virtual dataset from a manifest store would happen in the top-level function rather than the backends.
+
+```python
+ def open_virtual_dataset(
+    file_url: str,
+    object_reader: ObjectStore,
+    parser: Parser,
+    *,
+    drop_variables: Iterable[str] | None = None,
+    loadable_variables: Iterable[str] | None = None,
+    decode_times: bool | None = None,
+    cftime_variables: Iterable[str] | None = None,
+    indexes: Mapping[str, Index] | None = None,
+) → Dataset
+    """
+    Method for creating virtual Xarray dataset, only implemented in top level API
+    """
+    manifest_store = parser(
+        file_url = file_url,
+        object_reader = object_reader,
+    )
+    ds = manifest_store.to_virtual_dataset(
+        loadable_variables=loadable_variables,
+        decode_times=decode_times,
+        indexes=indexes,
+    )
+    return ds
+```
+
+> **NOTE**: Here the construction of a virtual dataset from a manifest store would happen in the top-level function rather than the backends.
+
+```python
+ def open_virtual_mfdataset(
+    paths: str | os.PathLike | Sequence[str | os.PathLike] | "NestedSequence[str | os.PathLike]",
+    object_reader: ObjectStore,
+    parser: Parser,
+    *,
+    concat_dim: (
+        str
+        | DataArray
+        | Index
+        | Sequence[str]
+        | Sequence[DataArray]
+        | Sequence[Index]
+        | None
+    ) = None,
+    compat: "CompatOptions" = "no_conflicts",
+    preprocess: Callable[[Dataset], Dataset] | None = None,
+    data_vars: Literal["all", "minimal", "different"] | list[str] = "all",
+    coords="different",
+    combine: Literal["by_coords", "nested"] = "by_coords",
+    parallel: Literal["dask", "lithops", False] | Executor = False,
+    join: "JoinOptions" = "outer",
+    attrs_file: str | os.PathLike | None = None,
+    combine_attrs: "CombineAttrsOptions" = "override",
+    **kwargs,
```

Comment:
@TomNicholas the lack of guessing functionality means that all inputs to `open_virtual_mfdataset` would need to use the same `ObjectReader` and `Parser`.